ArduPlane/commands_logic.cpp

```diff
@@ -526,17 +526,33 @@ bool Plane::verify_takeoff()
             // course. This keeps wings level until we are ready to
             // rotate, and also allows us to cope with arbitrary
             // compass errors for auto takeoff
-            float takeoff_course = wrap_PI(radians(gps.ground_course_cd()*0.01f)) - steer_state.locked_course_err;
-            takeoff_course = wrap_PI(takeoff_course);
-            steer_state.hold_course_cd = wrap_360_cd(degrees(takeoff_course)*100);
+            const float ground_course_rad = wrap_PI(ahrs.groundspeed_vector().angle());
+            const float ground_course = wrap_360(degrees(ground_course_rad));
+            const float takeoff_course_rad = wrap_PI(ground_course_rad - steer_state.locked_course_err);
+
+            steer_state.hold_course_cd = wrap_360_cd(degrees(takeoff_course_rad)*100);
+
+            // when true, we will follow the recorded takeoff heading.
+            // when false, we xtrack to a virtual waypoint 10km ahead of us on this takeoff heading
+            // TODO: turn this hardcoded value into a param/option?
+            steer_state.locked_course = false;
```

Comment:
An improvement would be to start off as true and then flip to false one we gain a little altitude, like 10m. That would solve any ground-steering vehicle issues and not induce any funny low-altitude behavior.

Reply:
This relies on starting with a good initial velocity vector. I had to look at this for a fast plane that had a longer takeoff run a while ago and ended up adding an option to use the takeoff waypoint location to set the desired ground track - see https://github.com/priseborough/ardupilot/commit/a6d2b2b68b4f5adae4e4c5ecc9a278e65c234127